### ISSUE TITLE
[Suggestion] Consider ObjectPool's ArrayBlockingQueue replacement

### ISSUE THREAD

**Issue by guidomedina** [romix/akka-kryo-serialization#78]
Have you considered changing the `ObjectPool` based on `ArrayBlockingQueue` for a better queue with MPSC or MPMC behavior?, examples below:

**MpscArrayQueue (for static size):** https://github.com/JCTools/JCTools/blob/master/jctools-core/src/main/java/org/jctools/queues/MpscArrayQueue.java
**MpscChunkedArrayQueue (for growing/shrinking size in chunks, linked list of arrays):** https://github.com/JCTools/JCTools/blob/master/jctools-core/src/main/java/org/jctools/queues/MpscChunkedArrayQueue.java

**Note:** If ArrayBlockingQueue is not much slower than these then it might not worth the effort, just pointing out these nice queues.

**Comment by luben** [comment]
Hi,

The last versions of Kryo already provide an obect pool. My intention is to rip-off our implementation and use the provided one. My main motivation is simplification of the code base. I don't think it will have noticable impact on the performance, kryo's pool is based on ConcurrentLinkedQueue

**Comment by guidomedina** [comment]
Hmmm, linked queues are bad news for many small messages, I would strongly consider modifying the one at Kryo to use a well known lock free algorithms, they are faster than ConcurrentLinkedQueue of course and no impact on GC.

You can even see that Netty 4 decided to use the MpscArrayQueue due to too many GC done at Netty 3 (One of the reasons Akka remote is so slow for high frequency messages)

**Comment by luben** [comment]
Yes, I will benchmark. We will need MPMC queue for our use case.

**Comment by guidomedina** [comment]
Here is your friend ;-)
https://github.com/JCTools/JCTools/blob/master/jctools-core/src/main/java/org/jctools/queues/MpmcArrayQueue.java

**Comment by guidomedina** [comment]
And yes I completely agree, but I would kill both birds with 1 shot, MPMC at Kryo and remove `ObjectPool` from akka-kryo

**Comment by luben** [comment]
Yes, that will be the ideal outcome, I will ask Kryo if thet are interested.

**Comment by guidomedina** [comment]
@luben for what I'm seeing it is too difficult to change the default Queue inside Kryo so I'm wondering if I can call Kryo and pass my queue, I see there is a builder, and then have akka-kryo use the same passed queue via implementation or default to their default.

That way I can just pass my MpmcArrayQueue with a proper size for my application.

**Comment by luben** [comment]
I think it's a good plan:

1. Replace the akka-kryo-serialization pool with the kryo pool
2. Make possible to configure a custom queue builder that will pass a queue instance to the kryo pool.

Is this what you are proposing?

This month I don't have free cycles (I am traverlling) but I can get to in early April unless you send a PR.

**Comment by guidomedina** [comment]
Yes that's what I'm proposing but my Scala is not that great, I can try though, also there is the blocking ObjectPool vs non-blocking KryoPool issue which will change the current contract.

**Comment by guidomedina** [comment]
@luben @romix I give up, I have spent one hour trying to code in Scala but unfortunately I was unable to have enough understanding **_-as a Java developer-_** to replace ObjectPool with KryoPool, I feel like an idiot because this task is a 5 minutes task for a Scala developer with few caveats:

1. Convert the current internal getInstance() into a KryoFactory.
2. Delete ObjectPool and replace it with KryoPool with the new KryoFactory.
3. Somehow allow the config to have an optional queue instance factory; I don't know Scala so wouldn't know what's easier if to pass an instance of a queue or parameters to create a queue (QueueFactory?) because some queues will require a size.

**Comment by luben** [comment]
No problem, I will get to it in 2-3 weeks.

**Comment by magro** [comment]
Btw, in the related kryo issue I proposed [how to do it here](https://github.com/EsotericSoftware/kryo/issues/403#issuecomment-194512872) (see also comments below).

**Comment by luben** [comment]
@magro thanks, I followed also the related discussion. Yes, may be providing a pool builder as you suggest will bring more flexibility.

**Comment by guidomedina** [comment]
I spent the whole day doing many changes but got to a dead-end.
Basically blocking and atomic size are sort of show stopper for whatever it is to be done later, with my limited Scala knowledge this is the best I could do for the next days until the next version is released, I have this deployed on our Archiva and the system running with it:

```
class ObjectPool[T](number: Int, newInstance: () => T) {

  private val pool = new MpmcArrayQueue[T](number)

  def fetch(): T = {
    pool.poll() match {
      case o if o != null => o
      case null => newInstance()
    }
  }

  def release(o: T): Unit = {
    pool.offer(o)
  }

  def add(o: T): Unit = {
    pool.add(o)
  }
}
```

**Comment by guidomedina** [comment]
@magro @luben 

A happier day, I have successfully coded the changes in Scala, just few caveats:

- Renamed ObjectPool to SerializerPool.
- Added a trait QueueCreator with a contract: `createQueue()`
- Cannot use KryoPool because you are not actually pooling Kryo instances but akka Serializer instances.
- New property for Kryo name custom-queue-creator, an optional property that defaults to a creator that returns a `ConcurrentLinkedQueue`
- poolSize property ceased to exist.
- Formatted and eliminated most IntelliJ/Scala warnings.
- Working on the documentation at the moment.

**Comment by guidomedina** [comment]
@luben @romix @magro please take a look at commit:

https://github.com/guidomedina/akka-kryo-serialization/commit/ed269a8d2b7b7eb0f04869325f5b147f7b0911a9

**Comment by guidomedina** [comment]
Re-pushed, eliminated creation of anonymous creator, instead let pool just grab a CLQ else the queue from creator.

**Comment by guidomedina** [comment]
Renamed `QueueCreator` to `QueueBuilder` and method `createQueue()` to `build()`

**Comment by luben** [comment]
LGTM, Yes, we may want to keep our akka serializer pool instead of using the kryo pool.

@romix, do you have any concerns?

**Comment by guidomedina** [comment]
@luben @romix I just force pushed again, IntelliJ was screwing up with formatting adding too many changes so I re-pushed, should be easier to check changes now.

**Comment by guidomedina** [comment]
@romix @luben two new PRs created; #80 & #81

**Comment by guidomedina** [comment]
Closed by #81
